packages/react/src/river/RiverAccordion/RiverAccordion.tsx

```diff
@@ -0,0 +1,227 @@
+  const {openIndex} = useRiverAccordionContext()
+  const panelId = useId()
+  const isOpen = index === openIndex
+
+  const itemContextValue = useMemo(
+    () => ({
+      id: panelId,
+      // Index optional for consumers, but always provided by parent
+      index: index as number,
+      isOpen,
+    }),
+    [panelId, index, isOpen],
+  )
+
+  const {heading, content, visual} = useMemo(() => {
+    return React.Children.toArray(children).reduce<{
+      heading: React.ReactElement<RiverAccordionHeadingProps> | null
+      content: React.ReactElement<RiverAccordionContentProps> | null
+      visual: React.ReactElement<RiverAccordionVisualProps> | null
+    }>(
+      (acc, child) => {
+        if (isRiverAccordionHeading(child)) {
+          acc.heading = child
+        }
+
+        if (isRiverAccordionContent(child)) {
+          acc.content = child
+        }
+
+        if (isRiverAccordionVisual(child)) {
+          acc.visual = child
+        }
+
+        return acc
+      },
+      {heading: null, content: null, visual: null},
+    )
+  }, [children])
+
+  if (!heading || !content || !visual) {
+    return null
+  }
+
+  return (
+    <RiverAccordionItemContext.Provider value={itemContextValue}>
+      <div
+        className={clsx(styles.RiverAccordion__item, isOpen && styles['RiverAccordion__item--open'], className)}
+        {...props}
+      >
+        {heading}
+        <div
+          className={styles.RiverAccordion__panel}
+          id={panelId}
+          aria-hidden={!isOpen}
+          {...(!isOpen && {inert: 'true'})}
+        >
+          {content}
+          {visual}
+        </div>
+      </div>
+    </RiverAccordionItemContext.Provider>
+  )
+}
+
+const RiverAccordionHeading = ({className, children, ...props}: RiverAccordionHeadingProps) => {
+  const {setOpenIndex} = useRiverAccordionContext()
+  const {id, index, isOpen} = useRiverAccordionItemContext()
+
+  const onClick = useCallback(() => {
+    if (!isOpen) {
+      setOpenIndex(index)
+    }
+  }, [index, isOpen, setOpenIndex])
+
+  return (
+    <Heading size="6" as="h3" className={clsx(styles.RiverAccordion__heading, className)} {...props}>
+      <button
+        type="button"
+        className={styles.RiverAccordion__trigger}
+        onClick={onClick}
+        aria-disabled={isOpen}
+        aria-controls={id}
+        aria-expanded={isOpen}
+      >
+        {children}
+        <PlusIcon size={24} className={styles.RiverAccordion__icon} />
+      </button>
+    </Heading>
+  )
+}
+
+const RiverAccordionContent = ({className, ...props}: RiverAccordionContentProps) => {
+  return <div className={clsx(styles.RiverAccordion__content, className)} {...props} />
+}
+
+const RiverAccordionVisual = ({className, ...props}: RiverAccordionVisualProps) => {
+  return <div className={clsx(styles.RiverAccordion__visual, className)} {...props} />
+}
+
+const createComponentTypeGuard =
```

Comment:
Nice!

Reply:
I think we have a helpers function, but you can move it there after. It's only worthwhile if it's a direct swap out in other components, so could do in a follow up.